render/shm_allocator.c

```diff
@@ -0,0 +1,98 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <wlr/util/log.h>
+#include "render/shm_allocator.h"
+#include "util/shm.h"
+
+static const struct wlr_buffer_impl buffer_impl;
+
+static struct wlr_shm_buffer *shm_buffer_from_buffer(
+		struct wlr_buffer *wlr_buffer) {
+	assert(wlr_buffer->impl == &buffer_impl);
+	return (struct wlr_shm_buffer *)wlr_buffer;
+}
+
+static void buffer_destroy(struct wlr_buffer *wlr_buffer) {
+	struct wlr_shm_buffer *buffer = shm_buffer_from_buffer(wlr_buffer);
+	munmap(buffer->data, buffer->size);
+	close(buffer->shm.fd);
+	free(buffer);
+}
+
+static bool buffer_get_shm(struct wlr_buffer *wlr_buffer,
+		struct wlr_shm_attributes *shm) {
+	struct wlr_shm_buffer *buffer = shm_buffer_from_buffer(wlr_buffer);
+	memcpy(shm, &buffer->shm, sizeof(*shm));
+	return true;
+}
+
+static bool buffer_get_data_ptr(struct wlr_buffer *wlr_buffer, void **data,
+		size_t *size) {
+	struct wlr_shm_buffer *buffer = shm_buffer_from_buffer(wlr_buffer);
+	*data = buffer->data;
+	*size = buffer->size;
+	return true;
+}
+
+static const struct wlr_buffer_impl buffer_impl = {
+	.destroy = buffer_destroy,
+	.get_shm = buffer_get_shm,
+	.get_data_ptr = buffer_get_data_ptr,
+};
+
+static struct wlr_buffer *allocator_create_buffer(
+		struct wlr_allocator *wlr_allocator, int width, int height,
+		const struct wlr_drm_format *format) {
+	struct wlr_shm_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		return NULL;
+	}
+	wlr_buffer_init(&buffer->base, &buffer_impl, width, height);
+
+	// TODO: consider using a single file for multiple buffers
+	int stride = width; // TODO: align
```

Comment:
stride should be `width * 4` as long as the bpp is 32 bits

Reply:
Good catch! Fixed.